_posts/2024-04-10-benchmarks-over-time.md

```diff
@@ -0,0 +1,547 @@
+            Plotly.plot( perf_over_time_window_area, parsed_json.data, parsed_json.layout );
+            });
+</script>
+
+Over the time horizon studied, window functions have **improved a dramatic 25x**!
+
+Window function performance was improved substantially with the 0.9.0 release in September 2023. 
+[14 different performance optimizations contributed](https://github.com/duckdb/duckdb/issues/7809#issuecomment-1679387022).
+Aggregate computation was vectorized (with special focus on the [segment tree data structure](https://www.vldb.org/pvldb/vol8/p1058-leis.pdf)).
+Work stealing enabled multi-threaded processing and sorting was adapted to run in parallel.
+Care was also taken to pre-allocate memory in larger batches. 
+
+DuckDB's window functions are also capable of processing larger than memory datasets. 
+We leave benchmarking that feature for future work!
+
+### Export
+
+<div id="perf_over_time_export_area" style="width:100%;height:400px;"></div>
+<script>
+    fetch('{{ site.baseurl }}/data/perf_over_time_export_area.json')
+        .then(res => res.json())
+        .then(parsed_json => {
+            let perf_over_time_export_area = document.getElementById('perf_over_time_export_area');
+            Plotly.plot( perf_over_time_export_area, parsed_json.data, parsed_json.layout );
+            });
+</script>
+
+Often DuckDB is not the final step in a workflow, so export performance has an impact. 
+Exports are **10x faster now!**
+Until recently, the DuckDB format was not backward compatible, so the recommended long term persistence format was Parquet.
+Parquet is also critical to interoperability with many other systems, especially data lakes. 
+DuckDB works well as a workflow engine, so exporting to other in memory formats is quite common as well.
+
+In version 0.5.1 we see significant improvements driven by switching from Pandas to Apache Arrow as the recommended in-memory export format.
+DuckDB's underlying data types share many similarities with Arrow, so data transfer is quite quick.
+
+Parquet export performance has improved by 4-5x over the course of the benchmark, with dramatic improvements in versions 0.8.1 and 0.10.2. 
+Version 0.8.1 added [parallel parquet writing](https://github.com/duckdb/duckdb/pull/7375) while continuing to preserve insertion order.
+
+The change driving the improvement in 0.10.2 was more subtle.
+When exporting strings with high cardinality, DuckDB decides whether or not to do dictionary compression depending on if it reduces file size. 
+In 0.10.2 onward, the [compression ratio is tested after a sample of the values are inserted into the dictionary](https://github.com/duckdb/duckdb/pull/11461), rather than after all values are added.
+This prevents substantial unnecessary processing for high cardinality columns where dictionary compression is unhelpful.
+
+#### Exporting Apache Arrow vs. Pandas vs. Parquet
+
+<div id="perf_over_time_export_arrow_pandas_parquet" style="width:100%;height:400px;"></div>
```

Comment:
In all of the plots except for the specific Arrow/Pandas/Parquet plots, I have a cutoff point (version 0.5.1) where before that, I use Pandas and afterwards I use Arrow. That was to try and mirror the real world where folks probably wouldn't move over to Arrow until it was better, and partially to avoid having that big spike in the summary graphs. 

So the purpose of that plot is to include that big spike and show when Arrow starts to beat Pandas. 

That sound ok, or still not something you want included?

Reply:
oh, looking more closely, I think that may just be a mistake on my side. Let me double check.

Reply:
I tried to take the cutoff into consideration but even that didn't explain it.

Reply:
Ok, I double checked, and the differences are:
* That 0.5.1 Pandas/Arrow cutoff 
* Area vs. line like you mentioned
* The line plot uses seconds instead of a relative measure, so that it is easier to add/remove series on

I would like to keep it in if possible since I think it tells an interesting story. Are there changes you would like me to make to clarify things a bit?

Also, totally agree on the placeholder - I will fix that now!

Reply:
I still do not understand the two plots in my [original comment](https://github.com/duckdb/duckdb-web/pull/3146#discussion_r1650371546), and in particular I don't understand how the cutoff works. If we change from Pandas to Arrow at version 0.5.1, than the starting point of the area "Export group by results to Arrow" in the top diagram should be 0.5.1. But it stretches across the entire version timeline, including versions prior to 0.5.1.

Reply:
Placeholders fixed on those 2 Pandas/Arrow/Parquet plots!

Reply:
Gabor had some great comments, so I've added placeholder zeroes in the area charts (so that when you filter on just 1 series, you can see the correct date axis), and I added in some explanations for the line charts to explain why they are different. Thanks!